File: MetroTtsEngine.kt
package com.metroreader.app.tts

import android.content.Context
import android.os.Bundle
import android.speech.tts.TextToSpeech
import android.speech.tts.UtteranceProgressListener
import android.speech.tts.Voice
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import java.util.Locale
import javax.inject.Inject
import javax.inject.Singleton

data class TtsState(
    val isInitialized: Boolean = false,
    val isPlaying: Boolean = false,
    val isPaused: Boolean = false,
    val currentChapterIndex: Int = 0,
    val currentSentenceIndex: Int = 0,
    val totalSentences: Int = 0,
    val speed: Float = 1.0f,
    val pitch: Float = 1.0f,
    val selectedVoice: Voice? = null,
    val availableVoices: List<Voice> = emptyList(),
    val error: String? = null,
)

/**
 * Core TTS engine wrapping Android TextToSpeech API.
 * Provides sentence-level playback with progress callbacks.
 */
@Singleton
class MetroTtsEngine @Inject constructor() {

    private var tts: TextToSpeech? = null
    private var sentences: List<String> = emptyList()
    private var chapterIndex: Int = 0
    private var sentenceIndex: Int = 0

    private val _state = MutableStateFlow(TtsState())
    val state: StateFlow<TtsState> = _state.asStateFlow()

    var onSentenceStart: ((chapterIndex: Int, sentenceIndex: Int) -> Unit)? = null
    var onSentenceDone: ((chapterIndex: Int, sentenceIndex: Int) -> Unit)? = null
    var onChapterDone: ((chapterIndex: Int) -> Unit)? = null
    var onPlaybackDone: (() -> Unit)? = null

    fun initialize(context: Context) {
        tts = TextToSpeech(context) { status ->
            if (status == TextToSpeech.SUCCESS) {
                val ttsInstance = tts ?: return@TextToSpeech
                ttsInstance.language = Locale.getDefault()

                val voices = ttsInstance.voices?.toList() ?: emptyList()
                _state.value = _state.value.copy(
                    isInitialized = true,
                    availableVoices = voices,
                )

                ttsInstance.setOnUtteranceProgressListener(object : UtteranceProgressListener() {
                    override fun onStart(utteranceId: String) {
                        val parts = utteranceId.split("_")
                        if (parts.size == 3 && parts[0] == "sentence") {
                            val ch = parts[1].toIntOrNull() ?: return
                            val si = parts[2].toIntOrNull() ?: return
                            sentenceIndex = si
                            _state.value = _state.value.copy(
                                currentChapterIndex = ch,
                                currentSentenceIndex = si,
                                isPlaying = true,
                            )
                            onSentenceStart?.invoke(ch, si)
                        }
                    }

                    override fun onDone(utteranceId: String) {
                        val parts = utteranceId.split("_")
                        if (parts.size == 3 && parts[0] == "sentence") {
                            val ch = parts[1].toIntOrNull() ?: return
                            val si = parts[2].toIntOrNull() ?: return
                            onSentenceDone?.invoke(ch, si)

                            // Queue next sentence
                            val nextIndex = si + 1
                            if (nextIndex < sentences.size) {
                                speakSentence(ch, nextIndex)
                            } else {
                                // Chapter done
                                _state.value = _state.value.copy(isPlaying = false)
                                onChapterDone?.invoke(ch)
                                onPlaybackDone?.invoke()
                            }
                        }
                    }

                    override fun onError(utteranceId: String) {
                        _state.value = _state.value.copy(
                            isPlaying = false,
                            error = "TTS error for utterance: $utteranceId"
                        )
                    }
                })
            } else {
                _state.value = _state.value.copy(error = "TTS initialization failed")
            }
        }
    }

    fun loadChapter(chapterIdx: Int, sentenceList: List<String>, startSentence: Int = 0) {
        stop()
        chapterIndex = chapterIdx
        sentences = sentenceList
        sentenceIndex = startSentence
        _state.value = _state.value.copy(
            currentChapterIndex = chapterIdx,
            currentSentenceIndex = startSentence,
            totalSentences = sentenceList.size,
        )
    }

    fun play() {
        if (!_state.value.isInitialized) return
        if (_state.value.isPaused) {
            // Resume from current position
            speakSentence(chapterIndex, sentenceIndex)
            _state.value = _state.value.copy(isPlaying = true, isPaused = false)
        } else {
            speakSentence(chapterIndex, sentenceIndex)
            _state.value = _state.value.copy(isPlaying = true)
        }
    }

    fun pause() {
        tts?.stop()
        _state.value = _state.value.copy(isPlaying = false, isPaused = true)
    }

    fun resume() = play()

    fun stop() {
        tts?.stop()
        sentenceIndex = 0
        _state.value = _state.value.copy(
            isPlaying = false,
            isPaused = false,
            currentSentenceIndex = 0,
        )
    }

    fun seekToSentence(index: Int) {
        tts?.stop()
        sentenceIndex = index.coerceIn(0, sentences.size - 1)
        _state.value = _state.value.copy(currentSentenceIndex = sentenceIndex)
        if (_state.value.isPlaying || _state.value.isPaused) {
            speakSentence(chapterIndex, sentenceIndex)
            _state.value = _state.value.copy(isPlaying = true, isPaused = false)
        }
    }

    fun setSpeed(speed: Float) {
        tts?.setSpeechRate(speed)
        _state.value = _state.value.copy(speed = speed)
    }

    fun setPitch(pitch: Float) {
        tts?.setPitch(pitch)
        _state.value = _state.value.copy(pitch = pitch)
    }

    fun setVoice(voice: Voice) {
        tts?.voice = voice
        _state.value = _state.value.copy(selectedVoice = voice)
    }

    fun getAvailableVoices(): List<Voice> = tts?.voices?.toList() ?: emptyList()

    private fun speakSentence(chapter: Int, index: Int) {
        val text = sentences.getOrNull(index) ?: return
        val params = Bundle()
        params.putString(TextToSpeech.Engine.KEY_PARAM_UTTERANCE_ID, "sentence_${chapter}_$index")
        tts?.speak(text, TextToSpeech.QUEUE_FLUSH, params, "sentence_${chapter}_$index")
    }

    fun shutdown() {
        tts?.stop()
        tts?.shutdown()
        tts = null
        _state.value = TtsState()
    }
}

File: MetroTTSEngine.swift
// MetroTTSEngine.swift
// Metro Reader — TTS Engine
// AVSpeechSynthesizer + Apple Neural Voices + sentence-level synchronization

import AVFoundation
import MediaPlayer
import Combine

// MARK: - TTS Engine State
enum TTSState: Equatable {
    case idle
    case loading
    case playing
    case paused
    case finished
}

// MARK: - MetroTTSEngine
@MainActor
final class MetroTTSEngine: NSObject, ObservableObject {

    // MARK: - Published State
    @Published var state: TTSState = .idle
    @Published var currentSentenceIndex: Int = -1
    @Published var currentChapterIndex: Int = 0
    @Published var speed: Float = 1.0
    @Published var pitch: Float = 1.0
    @Published var selectedVoice: AVSpeechSynthesisVoice? = nil
    @Published var availableVoices: [AVSpeechSynthesisVoice] = []

    // MARK: - Private
    private let synthesizer = AVSpeechSynthesizer()
    private var sentences: [String] = []
    private var chapters: [Chapter] = []
    private var pendingSentenceIndex: Int = 0
    private var bookTitle: String = ""
    private var bookAuthor: String = ""
    private var coverData: Data? = nil

    // MARK: - Callbacks
    var onSentenceHighlight: ((Int) -> Void)?
    var onChapterChange: ((Int) -> Void)?
    var onFinished: (() -> Void)?

    override init() {
        super.init()
        synthesizer.delegate = self
        loadAvailableVoices()
        setupRemoteCommandCenter()
    }

    // MARK: - Setup
    private func loadAvailableVoices() {
        availableVoices = AVSpeechSynthesisVoice.speechVoices()
            .filter { $0.language.hasPrefix("en") }
            .sorted { $0.name < $1.name }
        // Prefer enhanced/neural voices
        selectedVoice = availableVoices.first(where: { $0.quality == .enhanced || $0.quality == .premium })
            ?? availableVoices.first
    }

    // MARK: - Public API

    func load(book: Book, content: BookContent, startChapter: Int = 0, startSentence: Int = 0) {
        stop()
        chapters = content.chapters
        bookTitle = book.title
        bookAuthor = book.author
        coverData = book.coverData
        currentChapterIndex = startChapter
        pendingSentenceIndex = startSentence
        loadChapter(currentChapterIndex)
        updateNowPlaying()
    }

    func play() {
        guard state != .playing else { return }
        if state == .paused {
            synthesizer.continueSpeaking()
            state = .playing
            return
        }
        speakNextSentence()
    }

    func pause() {
        guard state == .playing else { return }
        synthesizer.pauseSpeaking(at: .word)
        state = .paused
    }

    func stop() {
        synthesizer.stopSpeaking(at: .immediate)
        state = .idle
        currentSentenceIndex = -1
        onSentenceHighlight?(-1)
    }

    func seekToSentence(_ index: Int) {
        synthesizer.stopSpeaking(at: .immediate)
        pendingSentenceIndex = max(0, min(index, sentences.count - 1))
        if state == .playing || state == .paused {
            state = .playing
            speakNextSentence()
        }
    }

    func nextChapter() {
        guard currentChapterIndex + 1 < chapters.count else { return }
        currentChapterIndex += 1
        loadChapter(currentChapterIndex)
        onChapterChange?(currentChapterIndex)
        if state == .playing { speakNextSentence() }
    }

    func previousChapter() {
        guard currentChapterIndex > 0 else { return }
        currentChapterIndex -= 1
        loadChapter(currentChapterIndex)
        onChapterChange?(currentChapterIndex)
        if state == .playing { speakNextSentence() }
    }

    func setVoice(_ voice: AVSpeechSynthesisVoice) {
        selectedVoice = voice
    }

    func setSpeed(_ newSpeed: Float) {
        speed = max(AVSpeechUtteranceMinimumSpeechRate, min(AVSpeechUtteranceMaximumSpeechRate, newSpeed))
    }

    func setPitch(_ newPitch: Float) {
        pitch = max(0.5, min(2.0, newPitch))
    }

    // MARK: - Private

    private func loadChapter(_ index: Int) {
        guard index < chapters.count else { return }
        sentences = chapters[index].sentences
        pendingSentenceIndex = 0
        currentSentenceIndex = -1
    }

    private func speakNextSentence() {
        guard pendingSentenceIndex < sentences.count else {
            // Move to next chapter
            if currentChapterIndex + 1 < chapters.count {
                nextChapter()
            } else {
                state = .finished
                onFinished?()
                onSentenceHighlight?(-1)
            }
            return
        }

        let sentence = sentences[pendingSentenceIndex]
        let utterance = AVSpeechUtterance(string: sentence)
        utterance.rate = speed * AVSpeechUtteranceDefaultSpeechRate
        utterance.pitchMultiplier = pitch
        utterance.voice = selectedVoice
        utterance.preUtteranceDelay = 0.05
        utterance.postUtteranceDelay = 0.05

        currentSentenceIndex = pendingSentenceIndex
        onSentenceHighlight?(pendingSentenceIndex)
        pendingSentenceIndex += 1
        state = .playing

        synthesizer.speak(utterance)
    }

    // MARK: - Now Playing / Lock Screen
    private func updateNowPlaying() {
        var info: [String: Any] = [
            MPMediaItemPropertyTitle: bookTitle,
            MPMediaItemPropertyArtist: bookAuthor,
            MPNowPlayingInfoPropertyPlaybackRate: state == .playing ? Double(speed) : 0.0,
            MPMediaItemPropertyPlaybackDuration: Double(sentences.count * 5), // estimate
            MPNowPlayingInfoPropertyElapsedPlaybackTime: Double(currentSentenceIndex * 5),
        ]
        if let data = coverData, let image = UIImage(data: data) {
            info[MPMediaItemPropertyArtwork] = MPMediaItemArtwork(boundsSize: CGSize(width: 300, height: 300)) { _ in image }
        }
        MPNowPlayingInfoCenter.default().nowPlayingInfo = info
    }

    private func setupRemoteCommandCenter() {
        let center = MPRemoteCommandCenter.shared()

        center.playCommand.addTarget { [weak self] _ in
            self?.play()
            return .success
        }
        center.pauseCommand.addTarget { [weak self] _ in
            self?.pause()
            return .success
        }
        center.stopCommand.addTarget { [weak self] _ in
            self?.stop()
            return .success
        }
        center.nextTrackCommand.addTarget { [weak self] _ in
            self?.nextChapter()
            return .success
        }
        center.previousTrackCommand.addTarget { [weak self] _ in
            self?.previousChapter()
            return .success
        }
        center.changePlaybackRateCommand.supportedPlaybackRates = [0.5, 0.75, 1.0, 1.25, 1.5, 2.0]
        center.changePlaybackRateCommand.addTarget { [weak self] event in
            if let e = event as? MPChangePlaybackRateCommandEvent {
                self?.setSpeed(Float(e.playbackRate))
            }
            return .success
        }
    }
}

// MARK: - AVSpeechSynthesizerDelegate
extension MetroTTSEngine: AVSpeechSynthesizerDelegate {

    nonisolated func speechSynthesizer(_ synthesizer: AVSpeechSynthesizer, didFinish utterance: AVSpeechUtterance) {
        Task { @MainActor in
            guard self.state == .playing else { return }
            self.speakNextSentence()
            self.updateNowPlaying()
        }
    }

    nonisolated func speechSynthesizer(_ synthesizer: AVSpeechSynthesizer, didPause utterance: AVSpeechUtterance) {
        Task { @MainActor in self.state = .paused }
    }

    nonisolated func speechSynthesizer(_ synthesizer: AVSpeechSynthesizer, didContinue utterance: AVSpeechUtterance) {
        Task { @MainActor in self.state = .playing }
    }

    nonisolated func speechSynthesizer(_ synthesizer: AVSpeechSynthesizer,
                                       willSpeakRangeOfSpeechString characterRange: NSRange,
                                       utterance: AVSpeechUtterance) {
        // Word-level highlighting (optional enhancement)
    }
}
